src/PythonClient/src/quixstreams/dataframes/platforms/quix/api.py

```diff
@@ -0,0 +1,129 @@
+from io import BytesIO
+from os import environ
+from typing import Optional
+from urllib.parse import urljoin
+from zipfile import ZipFile
+
+import requests
+
+__all__ = ("QuixPortalApiService",)
+
+
+class QuixPortalApiService:
+    """
+    A light wrapper around the Quix Portal Api. If used in the Quix Platform, it will
+    use that workspaces auth token and portal endpoint, else you must provide it.
+
+    Function names closely reflect the respective API endpoint,
+    each starting with the method [GET, POST, etc] followed by the endpoint path.
+
+    Results will be returned in the form of request's Response.json(), unless something
+    else is required. Non-200's will raise exceptions.
+
+    See the swagger documentation for more info about the endpoints.
+    """
+
+    def __init__(
+        self,
+        portal_api: Optional[str] = None,
+        auth_token: Optional[str] = None,
+        api_version: Optional[str] = None,
+        default_workspace_id: Optional[str] = None,
+    ):
+        self._portal_api = portal_api or environ.get(self.QuixEnvironmentMap.portal_api)
+        self._auth_token = auth_token or environ.get(self.QuixEnvironmentMap.sdk_token)
+        self.default_workspace_id = default_workspace_id or environ.get(
+            self.QuixEnvironmentMap.workspace_id
+        )
+        self.api_version = api_version or "2.0"
+        self.session = self._init_session()
+
+    class MissingConnectionRequirements(Exception):
+        pass
+
+    class QuixEnvironmentMap:
+        sdk_token = "Quix__Sdk__Token"
+        portal_api = "Quix__Portal__Api"
+        workspace_id = "Quix__Workspace__Id"
+
+    class SessionWithUrlBase(requests.Session):
+        def __init__(self, url_base: str):
+            self.url_base = url_base
+            super().__init__()
+
+        def request(self, method, url, **kwargs):
+            return super().request(
+                method, urljoin(base=self.url_base, url=url), **kwargs
+            )
+
+    def _init_session(self) -> SessionWithUrlBase:
+        if not (self._portal_api and self._auth_token):
+            missing = filter(
+                None,
+                [self._portal_api or "portal_api", self._auth_token or "auth_token"],
+            )
+            raise self.MissingConnectionRequirements(
+                f"Using the API requires a portal and token. Missing: {missing}"
+            )
+        s = self.SessionWithUrlBase(self._portal_api)
+        s.hooks = {"response": lambda r, *args, **kwargs: r.raise_for_status()}
+        s.headers.update(
+            {
+                "X-Version": self.api_version,
+                "Authorization": f"Bearer {self._auth_token}",
+            }
+        )
+        return s
+
+    def get_workspace_certificate(self, workspace_id: Optional[str] = None) -> ZipFile:
```

Comment:
Returning the ZipFile here complicates the work for the downstream code a bit.

Instead, we could extract `ca.cert` from the archive right in this function using https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.open and return plain `bytes`.

It is a cleaner option, and it's easier for the downstream code to decide what to do with `bytes` than to manipulate the ZIP archive.

Reply:
Yeah, makes sense. Changed it =)